### PR TITLE
Add basic auth security configuration

### PR DESCRIPTION
**Pull request by devinsba**
I would like to add security onto our zipkin server so we can do away with our IP whitelists that we currently use. This is a small first step toward that. (Ideally I will have an OAuth type solution at some point)

This leverages only existing code that was already in the project so it should only grow the artifact by a few properties files and a pom.

Running the docker images with these changes:

```
$ docker run -it -p 9411:9411 -e ZIPKIN_BASIC_SECURITY_USERNAME=zipkin \
    -e ZIPKIN_BASIC_SECURITY_PASSWORD=password \
    -e JAVA_OPTS='-Dspring.profiles.active=security' zipkin:local
```

### CODE REVIEW

File: zipkin-server/src/main/resources/zipkin-server-shared.yml [hunk 1]
```diff
@@ -133,6 +133,21 @@ spring:
...
+      password: ${ZIPKIN_BASIC_SECURITY_PASSWORD:}
+
+management:
+  security:
+    basic:
+      enabled: ${ZIPKIN_BASIC_SECURITY:false}
+      user:
+        name: ${ZIPKIN_BASIC_SECURITY_USERNAME:}
+        password: ${ZIPKIN_BASIC_SECURITY_PASSWORD:}
+
```
Comment (devinsba) [inline zipkin-server/src/main/resources/zipkin-server-shared.yml]: By default there will be no change for users other than seeing the randomly generated UUID password that spring-security-starter creates no matter what

**Comment by devinsba** [comment]
EDIT: turns out the tests wanted me to not have the `security.user.*` properties even if they are empty, so I ended up switching to a profile anyway

**Comment by codefromthecrypt** [comment]
I like the idea of finding a lightweight manner to progress #782 (which I think has been discussed offline as well, including using either custom auto-config or nginx proxy for the same).

Until reading further, it seems that even shared password would have drawbacks, notably that no instrumentation projects support it. That means if it was turned on, you could no longer use any instrumentation projects who use http transport, right? There's also the matter of how reusable this is to others.. I can't pretend to know that, so we should just ping folks.

I don't know the dependencies implied by spring-boot-starter-security, so it is probably more than just adding one existing jar and a config file layer.. have you tried this way?

**Comment by codefromthecrypt** [comment]
ps I don't see any tests, but we'd check the impact to see if this affects both the api and the UI (presumably it affects both)

**Comment by devinsba** [comment]
For the question of it affecting both the API and UI, yes it does affect both.

On the question of the reporters working with this: Sleuth does support this out of the box with a single extra bean being required to add the `Authorization` header.

Honestly I'd prefer a pluggable interface for this that would allow us to create a custom authentication module that we could drop into the docker image and would just work. I just wanted to put something out there quick in order to get a conversation started. I can try to find time to prototype an authentication provider component and create a PR with that? Or I can put together a proposal issue and try to get feedback from those in the gitter room.

**Comment by codefromthecrypt** [comment]
as a conversation starter, I think it succeeded :) To continue it, I'd
probably ping folks on issues who have formerly mentioned auth
requirements. Sleuth is one of many libraries, so if it were me, I'd try to
see what folks from other languages are interested in, or how other similar
projects secure their http transports for telemetry data.

I'm glad you started this.. if you can keep it up, please summarize updates
to #782 <https://github.com/openzipkin/zipkin/issues/782> I'll make some
notes there, which might help.

**Comment by Coalla** [comment]
When do you suppose to merge this?

**Comment by codefromthecrypt** [comment]
I dont think anything has progressed. It is important we dont add
speculative features. For example, none of the open source zipkin reporters
would work out of the box as they dont expect to authenticate.

Besides comments above and others made in the past, we need to make sure
this isnt a "tease feature" (looks neat but no one in the ecosystem
supports it)

File: zipkin-server/src/main/resources/zipkin-server-security.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+security:
+  basic:
+    enabled: true
+  user:
+    name: ${ZIPKIN_BASIC_SECURITY_USERNAME:}
```
Comment (maryoush) [inline zipkin-server/src/main/resources/zipkin-server-security.yml:5]: would that support multiple credentials ?

**Comment by devinsba** [comment]
Closing as my requirements have changed and a solution that works for everyone is unlikely to come from this work

**Comment by codefromthecrypt** [comment]
Are you in a position to share your new requirements? In back of mind was
thinking we can re-eval some things about modularity when we go spring boot
2

**Comment by devinsba** [comment]
We no longer need to use the http collector so an auth proxy just works
without having to worry about the reporters

**Comment by codefromthecrypt** [comment]
also obviated by this workaround https://github.com/openzipkin/zipkin/issues/782#issuecomment-433306013

**Pull request closed without merging**